File: frontend/src/modeling/lib/texture-coords.ts
const textureCoordCache = new Map<number, Float64Array>();

export const getTextureCoordinates = (
  textureIndex: number,
  textureWidth: number
): Float64Array => {
  const key = textureIndex * 65536 + textureWidth;
  let cached = textureCoordCache.get(key);
  if (cached) return cached;
  const textureSize = 1.0 / textureWidth;
  const halfPixel = textureSize * 0.5;
  const u = (textureIndex % textureWidth) * textureSize + halfPixel;
  const v = Math.floor(textureIndex / textureWidth) * textureSize + halfPixel;
  const flippedV = 1.0 - v;
  cached = new Float64Array([u, flippedV, u, flippedV, u, flippedV, u, flippedV]);
  textureCoordCache.set(key, cached);
  return cached;
};

File: wasm/src/texture_coords.rs
pub fn get_texture_coordinates(
    texture_index: i32,
    texture_width: i32,
) -> [f32; 8] {
    let texture_size = 1.0_f32 / texture_width as f32;
    let half_pixel = texture_size * 0.5;
    let u = (texture_index % texture_width) as f32 * texture_size + half_pixel;
    let v = (texture_index / texture_width) as f32 * texture_size + half_pixel;
    let flipped_v = 1.0_f32 - v;
    [u, flipped_v, u, flipped_v, u, flipped_v, u, flipped_v]
}
